src/core/backbone_router/local.cpp

```diff
@@ -0,0 +1,246 @@
+
+    if (aConfig.mMlrTimeout != mMlrTimeout)
+    {
+        mMlrTimeout = aConfig.mMlrTimeout;
+        update      = true;
+    }
+
+    if (aConfig.mSequenceNumber != mSequenceNumber)
+    {
+        mSequenceNumber = aConfig.mSequenceNumber;
+        update          = true;
+    }
+
+    if (update)
+    {
+        IgnoreReturnValue(AddService());
+    }
+
+    otLogDebgNetData("BBR local: seqno (%d), delay (%ds), timeout (%ds)", mSequenceNumber, mReregistrationDelay,
+                     mMlrTimeout);
+}
+
+otError Local::AddService(bool aForce)
+{
+    otError                               error       = OT_ERROR_INVALID_STATE;
+    uint8_t                               serviceData = NetworkData::ServiceTlv::kServiceDataBackboneRouter;
+    NetworkData::BackboneRouterServerData serverData;
+
+    VerifyOrExit(mState != OT_BACKBONE_ROUTER_STATE_DISABLED && Get<Mle::Mle>().IsAttached());
+
+    // If not by force, not register unless 1) there is no avilable Backbone Router service, or 2) it is PBBR itself
+    // refresh its dataset.
+    VerifyOrExit(aForce || !Get<BackboneRouter::Leader>().HasPrimary() || mState == OT_BACKBONE_ROUTER_STATE_PRIMARY);
+
+    serverData.SetSequenceNumber(mSequenceNumber);
+    serverData.SetReregistrationDelay(mReregistrationDelay);
+    serverData.SetMlrTimeout(mMlrTimeout);
+
+    error = Get<NetworkData::Local>().AddService(THREAD_ENTERPRISE_NUMBER, &serviceData, sizeof(serviceData), true,
+                                                 reinterpret_cast<const uint8_t *>(&serverData), sizeof(serverData));
+exit:
+    if (error == OT_ERROR_NONE)
+    {
+        mIsServiceAdded = true;
+        otLogInfoNetData("BBR Service added: seqno (%d), delay (%ds), timeout (%ds)", mSequenceNumber,
+                         mReregistrationDelay, mMlrTimeout);
+        error = Get<NetworkData::Local>().SendServerDataNotification();
+    }
+
+    otLogInfoNetData("Add BBR Service: %s", otThreadErrorToString(error));
+    return error;
+}
+
+otError Local::RemoveService(void)
+{
+    otError error       = OT_ERROR_NONE;
+    uint8_t serviceData = NetworkData::ServiceTlv::kServiceDataBackboneRouter;
+
+    error = Get<NetworkData::Local>().RemoveService(THREAD_ENTERPRISE_NUMBER, &serviceData, sizeof(serviceData));
+    SuccessOrExit(error);
+
+    error = Get<NetworkData::Local>().SendServerDataNotification();
+
+exit:
+    otLogInfoNetData("Remove BBR Service %s", otThreadErrorToString(error));
+    return error;
+}
+
+otError Local::AddBackboneRouterPrimaryAloc(void)
+{
+    memcpy(&mBackboneRouterPrimaryAloc.GetAddress(), &Get<Mle::MleRouter>().GetMeshLocal16(), 14);
+    mBackboneRouterPrimaryAloc.GetAddress().mFields.m16[7] = HostSwap16(Mle::kAloc16BackboneRouterPrimary);
+    return Get<ThreadNetif>().AddUnicastAddress(mBackboneRouterPrimaryAloc);
+}
+
+void Local::NotifyBackboneRouterPrimaryUpdate(Leader::State aState, const BackboneRouterConfig &aConfig)
+{
+    OT_UNUSED_VARIABLE(aConfig);
+
+    switch (aState)
```

Comment:
for states `kStateNone` and `kStateRemoved, should the PBBR `Reset()` to be SBBR?

Reply:
Agree that `kStateNone` & `PBBR` should never happen. 
But `kStateRemoved` and `PBBR` could happen. Thoughts?

Reply:
For what I can see, It may happen in some extreme scenario, for example, the non-leader PBBR could hear packets from the Thread Network, but any other devices in the Thread Network could not hear PBBR, then the Leader may remove the PBBR router id and remove its services. however in this case, PBBR itself should become detached when it detects its router id is not allocated any more in which it will reset local bbr state.
https://github.com/openthread/openthread/blob/adca16f505b568829c53ecbc774788d30a66ceed/src/core/thread/mle.cpp#L754

Any other possible such scenarios I missed?

Reply:
I was thinking from the standpoint of considering all possible logical cases. 

Could `Leader` switching cause some `Network Data` to be lost and led to this situation?

Reply:
If there is `new Leader` losing Network Data somehow, no matter whether the PBBR migrates to the new partition,  or forms new partition among with the devices in its communication range, or sticks to its old partition if `old leader` still in connection, it will either go through the same detach and attach process as above, or keep its primary state in old partition. 

Any more possible specific scenarios in your thoughts?